sql: fix attribution of multiple stages of physical plan

yuzefovich opened the pull request:
**physicalplan: ensure that ProcessorSpec.StageID is set everywhere**

This commit audits all places where we create `ProcessorSpec`s and
ensures that `StageID` field is set appropriately. 4 places are
modified, and for 3 of them this shouldn't have any effect (because we
have fresh plans with unset stage ID corresponding to reasonable "zero"
stage). The only place where this matters is `EnsureSingleStreamPerNode`
which is only used when planning set operations. This `StageID` value
only matters to how we visualize DistSQL diagrams (all processors of the
same stage show up at the same "level").

**sql: fix attribution of multiple stages of physical plan**

Our EXPLAIN infrastructure relies on `execNodeTraceMetadata` map that
maps a `planNode` to a list of `ComponentID`s that identify physical
plan components (i.e. processors) that correspond to the `planNode`.
Previously, we already handled the case when multiple processors
correspond to a single planNode (because the stage of the physical plan
is distributed), but we didn't handle the case when a single planNode
might result in multiple physical plan stages. As a result, we only used
the execution statistics of the _last_ stage that corresponds to the
planNode. Some example scenarios where this was problematic:
- distributed aggregation where we have "local" and "final" stages
- top K sort (we need a "final" stage to pick the true K rows from each
node)
- parallelized local scans (we need to merge the stream from each table
reader).

Concretely, this meant that we would get incomplete view of the
execution statistics (like missing SQL CPU time, or incorrect `sql
nodes` information).

This happened because we did the "association" in a single convenient
spot, but it was missing physical plan stages that were added internally
for some planNodes. This is now fixed by plumbing `finalizeLastStageCb`
that is invoked after a new physical plan stage is finalized.
Additionally, `execNodeTraceMetadata` has been adjusted to track
different components in stages too - this is needed since some of the
execution statistics (like row count and vectorized batch count) cannot
be aggregated across all stages.

Note that this change should not incur any allocations when we're not
collecting the execution statistics which is the case on the fast path.

Also this change is a bit unfortunate, but given that we'll need to
rethink how we do execution statistics for the spec factory I decided to
not make any major refactors here, and instead fix the issue in the most
simple way.

Although it is a bug fix, it seems like an edge case to me, so I decided
to omit the release note.

Fixes: #99741.
Fixes: #141530.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142292)
<!-- Reviewable:end -->

yuzefovich committed:
physicalplan: ensure that ProcessorSpec.StageID is set everywhere

This commit audits all places where we create `ProcessorSpec`s and
ensures that `StageID` field is set appropriately. 4 places are
modified, and for 3 of them this shouldn't have any effect (because we
have fresh plans with unset stage ID corresponding to reasonable "zero"
stage). The only place where this matters is `EnsureSingleStreamPerNode`
which is only used when planning set operations. This `StageID` value
only matters to how we visualize DistSQL diagrams (all processors of the
same stage show up at the same "level").

Release note: None

yuzefovich committed:
sql: fix attribution of multiple stages of physical plan

Our EXPLAIN infrastructure relies on `execNodeTraceMetadata` map that
maps a `planNode` to a list of `ComponentID`s that identify physical
plan components (i.e. processors) that correspond to the `planNode`.
Previously, we already handled the case when multiple processors
correspond to a single planNode (because the stage of the physical plan
is distributed), but we didn't handle the case when a single planNode
might result in multiple physical plan stages. As a result, we only used
the execution statistics of the _last_ stage that corresponds to the
planNode. Some example scenarios where this was problematic:
- distributed aggregation where we have "local" and "final" stages
- top K sort (we need a "final" stage to pick the true K rows from each
node)
- parallelized local scans (we need to merge the stream from each table
reader).

Concretely, this meant that we would get incomplete view of the
execution statistics (like missing SQL CPU time, or incorrect `sql
nodes` information).

This happened because we did the "association" in a single convenient
spot, but it was missing physical plan stages that were added internally
for some planNodes. This is now fixed by plumbing `finalizeLastStageCb`
that is invoked after a new physical plan stage is finalized.
Additionally, `execNodeTraceMetadata` has been adjusted to track
different components in stages too - this is needed since some of the
execution statistics (like row count and vectorized batch count) cannot
be aggregated across all stages.

Note that this change should not incur any allocations when we're not
collecting the execution statistics which is the case on the fast path.

Also this change is a bit unfortunate, but given that we'll need to
rethink how we do execution statistics for the spec factory I decided to
not make any major refactors here, and instead fix the issue in the most
simple way.

Although it is a bug fix, it seems like an edge case to me, so I decided
to omit the release note.

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 11 of 11 files at r1, 17 of 17 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142292#-:-OKZrcTN-IBQLk-C--HG:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Thanks for the quick review!

bors r+
